waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,475 @@
+    return err("Cannot establish sync connection")
+
+  let hashes: seq[WakuMessageHash] = (await self.request(conn)).valueOr:
+    debug "sync session ended",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+    return err("Sync request error: " & error)
+
+  return ok((hashes, peer))
+
+proc handleLoop(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?serverInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  var sent = initialized
+
+  while true:
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = received.payload
+
+    let reconciled = (?received.serverReconcile()).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      let hashes = await completed.serverTerminate()
+
+      return ok(hashes)
+
+    sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = reconciled.payload
+
+proc initProtocolHandler(self: WakuSync) =
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    debug "sync session requested",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+    let hashes = (await self.handleLoop(conn)).valueOr:
+      debug "sync session ended",
+        server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
+        error "transfer callback failed", error = $error
+
+    debug "sync session ended gracefully",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+  self.handler = handle
+  self.codec = WakuSyncCodec
+
+proc initPruningHandler(self: WakuSync, wakuArchive: WakuArchive) =
+  if wakuArchive.isNil():
+    error "waku archive unavailable"
+    return
+
+  self.pruneCallBack = some(
+    proc(
+        pruneStart: Timestamp, pruneStop: Timestamp, cursor: Option[WakuMessageHash]
+    ): Future[
+        Result[(seq[(WakuMessageHash, Timestamp)], Option[WakuMessageHash]), string]
+    ] {.async: (raises: []), closure.} =
+      let archiveCursor =
+        if cursor.isSome():
+          some(ArchiveCursor(hash: cursor.get()))
+        else:
+          none(ArchiveCursor)
+
+      let query = ArchiveQuery(
+        includeData: true,
+        cursor: archiveCursor,
+        startTime: some(pruneStart),
+        endTime: some(pruneStop),
+        pageSize: 100,
+      )
+
+      let catchable = catch:
+        await wakuArchive.findMessages(query)
```

Comment:
I agree it's not the best but it's fine for now. We plan on improving the Negentropy storage in the future so that querying the DB would not be necessary.